src/boot/bootctl-status.c

```diff
@@ -503,6 +504,224 @@ int verb_status(int argc, char *argv[], void *userdata) {
+        }
+}
+
+static int count_known_files(const BootConfig *config, const char* root, Hashmap **ret_known_files) {
+        _cleanup_(hashmap_free_free_keyp) Hashmap *known_files = NULL;
+        int r = 0;
+        assert(config);
+        assert(ret_known_files);
+
+        known_files = hashmap_new(&path_hash_ops);
+        if (!known_files)
+                return log_oom();
+
+        for (size_t i = 0; i < config->n_entries; i++) {
+                const BootEntry *e = config->entries + i;
+
+                if (!e->root || strcmp(e->root, root))
+                        continue;
+
+                r = ref_file(known_files, e->kernel, 1);
+                if (r < 0)
+                        return r;
+                STRV_FOREACH(s, e->initrd) {
+                        r = ref_file(known_files, *s, 1);
+                        if (r < 0)
+                                return r;
+                }
+                r = ref_file(known_files, e->device_tree, 1);
+                if (r < 0)
+                        return r;
+                STRV_FOREACH(s, e->device_tree_overlay) {
+                        r = ref_file(known_files, *s, 1);
+                        if (r < 0)
+                                return r;
+                }
+        }
+
+        *ret_known_files = TAKE_PTR(known_files);
+
+        return 0;
+}
+
+static int unlink_entry(const BootConfig *config, const char *root, const char *fn) {
+        _cleanup_(hashmap_free_free_keyp) Hashmap *known_files = NULL;
+        const BootEntry *e = NULL;
+        int r;
+
+        r = count_known_files(config, root, &known_files);
+        if (r < 0)
+                return log_error_errno(r, "Failed to count files in %s: %m", root);
+
+        for (size_t i = 0; i < config->n_entries; i++) {
+                const BootEntry *ei = &config->entries[i];
+
+                if (!ei->root || strcmp(ei->root, root))
+                        continue;
+                if (fnmatch(fn, ei->id, FNM_CASEFOLD))
+                        continue;
+
+                e = ei;
+                if (i == (size_t) config->default_entry)
+                        log_warning("%s is the default boot entry", fn);
+                if (i == (size_t) config->selected_entry)
+                        log_warning("%s is the selected boot entry", fn);
+                break;
+        }
+
+        if (!e)
+                return -ENOENT;
+
+        if (e->kernel) {
+                _cleanup_free_ char *d = NULL;
+
+                deref_unlink_file(known_files, e->kernel, e->root);
+                STRV_FOREACH(s, e->initrd)
+                        deref_unlink_file(known_files, *s, e->root);
+                deref_unlink_file(known_files, e->device_tree, e->root);
+                STRV_FOREACH(s, e->device_tree_overlay)
+                        deref_unlink_file(known_files, *s, e->root);
+
+                if (path_extract_directory(e->kernel, &d) == 0)
+                        (void) rmdir(d);
```

Comment:
please move this into deref_unlink_file() and do this for all files we remove here, not just the kernel. I mean, yeah this is quite likely going to see ENOTEMPTY on the first attempts, but that should be OK.

Reply:
also would be good to warn about errors here that are unexpected (i.e. any errors except for ENOTEMPTY i guess

Reply:
also use chase_symlinks_and_unlink() with AT_REMOVEDIR